packages/js/src/server/integrations/uncaught_exception.ts

```diff
@@ -25,49 +25,59 @@ function hasOtherUncaughtExceptionListeners() {
   return process.listeners('uncaughtException').length > 1
 }
 
+function handleUncaughtException(uncaughtError: Error, client: typeof Client) {
+  if (isReporting) { 
+    return 
+  }
+
+  // In the "load" function in the exported plugin below, we 
+  // never attach this listener if enableUncaught is false...
+  // so why do we need this?
+  if (!client.config.enableUncaught) {
+    client.config.afterUncaught(uncaughtError)
+    if (!hasOtherUncaughtExceptionListeners()) {
+      fatallyLogAndExit(uncaughtError)
+    }
+    return
+  }
+
+  // report only the first error - prevent reporting recursive errors
+  if (handlerAlreadyCalled) {
+    if (!hasOtherUncaughtExceptionListeners()) {
+      fatallyLogAndExit(uncaughtError)
+    }
+    return
+  }
+
+  isReporting = true
+  client.notify(uncaughtError, {
+    afterNotify: (_err, _notice) => {
+      isReporting = false
+      handlerAlreadyCalled = true
+      client.config.afterUncaught(uncaughtError)
+      if (!hasOtherUncaughtExceptionListeners()) {
+        fatallyLogAndExit(uncaughtError)
+      }
+    }
+  })
+}
+
 export default function (): Types.Plugin {
   return {
     load: (client: typeof Client) => {
       if (!client.config.enableUncaught) {
         return
       }
-
       removeAwsLambdaListener()
-
-
       process.on('uncaughtException', function honeybadgerUncaughtExceptionListener(uncaughtError) {
-        if (!client.config.enableUncaught) {
-          client.config.afterUncaught(uncaughtError)
-          if (!hasOtherUncaughtExceptionListeners()) {
-            fatallyLogAndExit(uncaughtError)
-          }
-          return
-        }
-
-        // report only the first error - prevent reporting recursive errors
-        if (handlerAlreadyCalled) {
-          if (!hasOtherUncaughtExceptionListeners()) {
-            fatallyLogAndExit(uncaughtError)
-          }
-          return
-        }
-
-        if (isReporting) {
-          return
-        }
-
-        isReporting = true
-        client.notify(uncaughtError, {
-          afterNotify: (_err, _notice) => {
-            isReporting = false
-            handlerAlreadyCalled = true
-            client.config.afterUncaught(uncaughtError)
-            if (!hasOtherUncaughtExceptionListeners()) {
-              fatallyLogAndExit(uncaughtError)
-            }
-          }
-        })
+        handleUncaughtException(uncaughtError, client)
       })
     }
   }
 }
+
+export const exportedForTesting = {
```

Comment:
Not everyone likes this way of getting access to functions for unit testing -- I'm open to other suggestions. 

`rewire` or another library like that is probably an option if folks don't like this, it's just more setup.

Reply:
🤔 Not a fan of this either, but I see why you have to do it. How about we just export these functions in the normal way?
```
export function handleUncaughtException() { ... }
export function setIsReporting() { ... }
export function setHandlerAlreadyCalled() { ... }
```
And you can use `setIsReporting`, `setHandlerAlreadyCalled` inside this file as well, instead of modifying it directly.

Reply:
In general it's not ideal to export anything that isn't intended to be used by other modules -- a later developer may think it's exported to be used in the code when it's actually just exported to be used in a test. For example, it could cause issues if someone used `setIsReporting` from some other file. 

In this case, I'm honestly not worried about it either way -- it's a small-ish package and I don't think there's likely to be confusion. Also if we are considering the switch to `uncaughtExceptionMonitor`, a lot of this will get re-worked.

Reply:
> In general it's not ideal to export anything that isn't intended to be used by other modules

I agree. But exporting here does not make these functions available to other modules.

> In this case, I'm honestly not worried about it either way -- it's a small-ish package and I don't think there's likely to be confusion.

My concern is that we don't set a precedence and start testing our code this way, i.e. exporting for the sake of tests. It makes me think that there is a better solution out there.
I see 2 possible solutions:
- use rewire
- rewrite this piece of logic so that it can be mocked, for example we could create a `UncaughtExceptionProcessor` class in a separate file wrapped with this logic. Then we can mock the entire file.

> Also if we are considering the switch to uncaughtExceptionMonitor, a lot of this will get re-worked.

True. But as your [investigation](https://github.com/honeybadger-io/honeybadger-js/issues/960#issuecomment-1675320288) has revealed, we may be sticking with this code after all 😅 .

Reply:
@subzero10 I went ahead and made a separate class for this -- I think that's a reasonable choice anyway since most of this package is structured as classes. I now have an integration test failure that I'm looking into (it appears unrelated at first glance).

Reply:
Thanks for doing that! The updated code seems much cleaner 👌